consensus/types.go

```diff
@@ -1,53 +1,168 @@
+//   - Aggregation of timeout messages for the current view
+//   - Signaling its caller when it can vote on the current chain tip
+//   - Signaling its caller when it has timed out the current view
+//   - Signaling its caller when it has a QC for the current block
+//   - Signaling its caller when it has a timeout QC for the current view
+//
+// When a new block is connected to the chain, the caller is expected to update the chain tip. The module
+// resets all internal data structures and timers to handle all of the above based on the new chain tip.
+//
+// This module is very simply and only houses the logic that decide what action to perform next given the
+// current chain tip. The module does not track the history of blocks, and instead needs its caller to
+// update the block at the current chain tip. It expects its caller to maintain the block chain,
+// the index of all past blocks, to perform QC validations for incoming blocks, to handle the commit rule,
+// and only then to pass the validated chain tip. Note: this module takes the provided chain tip as a
+// trusted input and does NOT validate any incoming blocks. This also mean the module expects its caller to
+// track historical vote and timeout messages it has sent so as to not vote more than once at a given view
+// or block height.
 type FastHotStuffConsensus struct {
-	nextBlockProposalTime time.Time
-	nextTimeoutTime       time.Time
+	lock sync.RWMutex
+
+	blockConstructionCadence time.Duration
+	timeoutBaseDuration      time.Duration
 
-	internalTimersUpdated chan interface{}
-	quit                  chan interface{}
+	nextBlockConstructionTime time.Time
+	nextTimeoutTime           time.Time
 
-	// votesSeen and timeoutsSeen are in-memory maps of all the votes and timeouts we've
-	// seen so far, organized by their block hash and then by the public key of the voter.
-	votesSeen    map[BlockHash]map[bls.PublicKey]*bls.Signature // TODO: this should be a vote message
-	timeoutsSeen map[BlockHash]map[bls.PublicKey]*bls.Signature // TODO: this should be a timeout message
+	// The latest block accepted by the caller. We only keep track of and build on top of the chain
+	// tip here. In the event of a fork, we expect the new tip to be resolved and provided by the caller
+	chainTip Block
+	// The current view at which we expect to see or propose the next block. In the event of a timeout,
+	// the timeout signal will be triggered for this view.
+	currentView uint64
+	// The validator set sorted in decreasing order of stake amount, with a consistent tie-breaking
+	// scheme. This validator set is expected to be valid for validating votes and timeouts for the
+	// next block height.
+	validators []Validator
 
+	// votesSeen is an in-memory map of all the votes we've seen so far, organized by the public key
+	// of the sender and the block hash of the block that was voted on.
+	votesSeen map[votesSeenMapKey]VoteMessage
```

Comment:
I was wrong about these - we want the nested maps. sorry

Reply:
Reverted.